KEYCLOAK-13160 Add support for Privacy Preserving Logging

thomasdarimont opened the pull request:
The PR provides a new `PrivacyFilterProvider` SPI, which allows applying custom filtering of log messages generated via Keycloak's `jboss-logging` EventListener. 

For a start we ship with two implementations: A default no-op `DefaultEmptyPrivacyFilterProvider`, and a configurable `AnonymizingPrivacyFilterProvider`. 
The latter uses a customizable `Anonymizer` to perform the obfuscation / masking  / filtering of PII information. 
Special Type-Hints allow to customize the obfuscation. 
Keycloak developers can now implement their own, `PrivacyFilterProvider` implementation or extend the `AnonymizingPrivacyFilterProvider`. 

It is also possible to use the `PrivacyFilterProvider` directly, e.g. for own log messages:
```java
keycloakSession.getProvider(PrivacyFilterProvider.class).filter("some data", PrivacyFilterProvider.PII);
```

It was also necessary to adapt the `BruteForceProtector` to use the JBoss-Logging EventListener to benefit from the new SPI.

You can find a [discussion about this on the keycloak-dev mailing list](https://groups.google.com/forum/#!topic/keycloak-dev/d43PoN7ZCkw). 

Anonymization examples for `username` / `emails` / `userId`:
- `user.unknown@emailprovider.com --> use%r.com`
- `valid.user@emailprovider.com --> val%r.com`
- `userId=45f31bdf-7afd-4d42-86ad-6b6c6e3eded3 --> 45f%ed3`
- `ipAddress= 127.0.0.1 -> 12%0.1`

With this in place, we can generate log messages that look like this:

```
17:26:30,732 WARN XNIO-1 task-1 [org.keycloak.events] type=LOGIN_ERROR, realmId=demo, clientId=null, userId=null, ipAddress=12%0.1, error=expired_code, restart_after_timeout=true, authSessionParentId=0a1a474e-c08d-466f-881f-efefbb788bb3, authSessionTabId=5OiP-XRunw4

17:26:33,863 WARN XNIO-1 task-2 [org.keycloak.events] type=LOGIN_ERROR, realmId=demo, clientId=account, userId=bf%9ab, ipAddress=12%0.1, error=invalid_user_credentials, auth_method=openid-connect, redirect_uri=http://0.0.0.0:8081/auth/realms/demo/account/login-redirect, code_id=0a1a474e-c08d-466f-881f-efefbb788bb3, username=te%ter, authSessionParentId=0a1a474e-c08d-466f-881f-efefbb788bb3, authSessionTabId=5OiP-XRunw4

17:26:33,869 WARN Brute Force Protector [org.keycloak.events] type=LOGIN_ERROR, realmId=demo, clientId=null, userId=bf%9ab, ipAddress=12%0.1, error=recording login failure for user
```

<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
KEYCLOAK-13160 Support GDPR compliant anonymization in Log Messages

- Made extraction of PII information in JBossLoggingEventListenerProvider
  customizable in subclasses.
- Introduced GdprAwareJBossLoggingEventListenerProvider
- Introduced Anonymizer interface with Default implementation
  which allows for customizable anonymization rules

thomasdarimont committed:
KEYCLOAK-13160 Support GDPR compliant anonymization in BruteForce Detector

- Revised Bruteforce detector to the Keycloak event subsystem instead
  of plain logging. This allows to anonymize PII in one spot.

thomasdarimont committed:
KEYCLOAK-13160 Change listener name to PrivacyPreservingJBossLoggingEventListenerProvider

thomasdarimont committed:
KEYCLOAK-13160 NOOP_CONNECTION -> NULL_CONNECTION

thomasdarimont committed:
KEYCLOAK-13160 Make JBossLoggingEventListenerProvider privacy aware

- Introduced PrivacyProvider SPI to developers to specify their own privacy filtering
- Introduced NO-OP implementation DefaultPrivacyProvider which is also the server default
- Introduced configurable AnonymizingPrivacyProvider
- JBossLoggingEventListenerProvider uses the configured PrivacyProvider to filter sensitive event message fields
- Removed PrivacyPreservingJBossLoggingEventListenerProvider implementation
- The anonymizer can be enabled via -Dkeycloak.privacy.provider=anonymize when running the KeycloakServer from the test-suite.
  Note: you might need to set the "success-level" log-level in JBossLoggingEventListenerProvider to info to see the result.

stianst commented on services/src/main/java/org/keycloak/privacy/PrivacyProvider.java:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @author <a href="mailto:thomas.darimont@googlemail.com">Thomas Darimont</a>
+ */
+public interface PrivacyProvider extends Provider {
```
Wonder if we should add "String filter (String input)", then have a default method for the current two methods in here that just uses that. That would make it simpler to implement one if you don't care about type, etc. and also means you can have a fallback if there's additional things added in the future.

thomasdarimont committed:
KEYCLOAK-13160 Refactoring

- Introduce single filter method for generic filtering of inputs
- Add support for default anonymizing of fields via "null" field name.

thomasdarimont committed:
KEYCLOAK-13160 Refactoring of PrivacyFilterProvider infrastructure

- Rename Privacy* to PrivacyFilter*
- Rename SPI ID from `privacy` to `privacy-filter`
- Introduce single filter method for generic filtering of inputs
- Add support for default anonymizing of fields via "null" field name.

thomasdarimont commented:
@stianst I adapted the PR as you requested, as I introduced a new single-arg filter Method to the `PrivacyProvider` interface. However I think most real-world implementations will probably override the other to method variants, which have explicit field type information. 

Besides that, there might also be implementations, which want to hash or encrypt PII inputs that might find the single argument method that handles all field inputs in a common way quite handy.

thomasdarimont committed:
KEYCLOAK-13160 Refactoring of PrivacyFilterProvider infrastructure

- Rename Privacy* to PrivacyFilter*
- Rename SPI ID from `privacy` to `privacy-filter`
- Introduce single filter method for generic filtering of inputs
- Add support for default anonymizing of fields via "null" field name.
- Removed superfluous `key` parameter   String filter(String field, String input, Event event)  of PrivacyFilterProvider

thomasdarimont commented:
Based on this, we could also provide a `PrivacyFilterProvider` implementation that supports encryption of PII values with a shared key (+unique salt per user). Only administrators, auditors or security personnel, would have access to the keys to be able to decrypt the actual information from the sensitive log fields. This model would also support to rotate keys and by simply throwing away the old encryption key, the older log information cannot be decrypted anymore to reveal to sensitive PII. However, those logs could still be used for analysis. It would be great if the log key rotation could  be done at runtime.

thomasdarimont committed:
KEYCLOAK-13160 Revise AnonymizingPrivacyFilterProviderFactory

thomasdarimont commented:
Just rebased this on top of current master.

thomasdarimont committed:
KEYCLOAK-13160 Refactored PrivacyFilterProvider infrastructure

thomasdarimont commented:
@stianst I just did a general overhaul of this PR. I think it is now ready for review.

Users can now configure custom field / type mappings that should be considered for anonymization.

I tested this with the `org.keycloak.testsuite.KeycloakServer` by configuring the log level of success-events in `org.keycloak.events.log.JBossLoggingEventListenerProviderFactory#init` to `info`.
Additionally I configured the `privacy-filter` SPI in `keycloak-server.json` to `anonymize`
```json
  "privacy-filter":{
         "provider": "${keycloak.privacyFilter.provider:anonymize}",
         "anonymize": {
         }
     },
```

With this in place, just login with a user and look at the logs.

A login for the user admin on localhost produces the following log messages (with succes-level: info)
```
00:29:46,940 INFO  XNIO-1 task-2 [org.keycloak.events] type=LOGIN, realmId=master, clientId=security-admin-console, userId=b3%29f, ipAddress=12%0.1, auth_method=openid-connect, auth_type=code, redirect_uri=http://0.0.0.0:8081/auth/admin/master/console/, consent=no_consent_required, code_id=0b027f92-e067-4fd2-bcdf-8cd9a23e5e37, username=admin, authSessionParentId=0b027f92-e067-4fd2-bcdf-8cd9a23e5e37, authSessionTabId=0TxG_4enp3I
00:29:47,550 INFO  XNIO-1 task-2 [org.keycloak.events] type=CODE_TO_TOKEN, realmId=master, clientId=security-admin-console, userId=b3%29f, ipAddress=12%0.1, token_id=2c2f568e-3d80-4eae-b979-482e8326edaa, grant_type=authorization_code, refresh_token_type=Refresh, scope='openid profile email', refresh_token_id=b8865fe4-fb8e-49d6-b2d9-f02913fd26be, code_id=0b027f92-e067-4fd2-bcdf-8cd9a23e5e37, client_auth_method=client-secret
```

thomasdarimont committed:
KEYCLOAK-13160 Revise documentation of AnonymizingPrivacyFilterProvider

thomasdarimont committed:
KEYCLOAK-13160 Fix javadoc of JBossLoggingEventListenerProvider

thomasdarimont committed:
KEYCLOAK-13160 Add tests for AnonymizingPrivacyFilterProvider

thomasdarimont committed:
KEYCLOAK-13160 Revise DefaultEmptyPrivacyFilterProviderFactory

thomasdarimont committed:
KEYCLOAK-13160 Refactor AnonymizingPrivacyFilterProviderFactory

thomasdarimont committed:
KEYCLOAK-13923 Make parsing of custom AnonymizingPrivacyFilterProviderFactory configuration more robust

thomasdarimont committed:
KEYCLOAK-13923 Move built-in type-hint constants into PrivacyTypeHints and add support for rendering input as plain

thomasdarimont commented:
I'll revise this PR with some inspirations from [SQL Servers dynamic data masking feature](https://docs.microsoft.com/en-us/sql/relational-databases/security/dynamic-data-masking?view=sql-server-ver15)

thomasdarimont commented:
I'm closing this PR and create a new one with a leaner implementation.

mazenaissa commented:
Hello @thomasdarimont,
Any update about this feature please?

SeJunB commented:
This feature would be nice. Until it's finished, our team is currently using a Quarkus Extension to filter/mask items in logs.
 See https://github.com/SeJunB/KeycloakLogFilter for an example.
